# 01 Importing libraries

In [144]:
# Import libraries 
import pandas as pd
import numpy as np
import os

# 02 Importing data

In [145]:
# Srigns for path
path=r'C:\Users\Oksana Stepanova\OneDrive\Документи\Instacart Basket Analysis'

In [146]:
# Import data Products
df_prods=pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'))

In [147]:
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [148]:
# Import data Orders
df_ords=pd.read_csv(os.path.join(path,'02 Data', 'Prepared Data', 'orders_wrangled.csv'))

In [149]:
df_ords.head()

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0           0   2539329        1             1                   2   
1           1   2398795        1             2                   3   
2           2    473747        1             3                   3   
3           3   2254736        1             4                   4   
4           4    431534        1             5                   4   

   order_hour_of_day  days_since_prior_order  
0                  8                     NaN  
1                  7                    15.0  
2                 12                    21.0  
3                  7                    29.0  
4                 15                    28.0

# 03 Mixed-type data

In [150]:
# Create a dataframe
df_test=pd.DataFrame()

In [151]:
# Create a mixed-type column
df_test['mix']=['a', 'b', 1, True]

In [152]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [153]:
# Check for mixed types
for col in df_test.columns.tolist():
    weird=(df_test[[col]].map(type) !=df_test[[col]].iloc[0].apply(type)).any(axis=1)
    if len (df_test[weird])>0:
        print(col)

mix


In [154]:
# Change data type in df_test
df_test['mix']=df_test['mix'].astype('str')

In [155]:
df_test.describe()

mix
count    4
unique   4
top      a
freq     1

In [156]:
df_test.dtypes

mix    object
dtype: object

# 04 Missing values

In [157]:
# Find missing values in Products
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [158]:
# Create Subset showing rows with missing product names
df_prods_nan=df_prods[df_prods['product_name'].isnull()==True]

In [159]:
df_prods_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [160]:
# Check the number of rows and columns in Products
df_prods.shape

(49693, 5)

In [161]:
# Create new df of Products and remove missing values in product_name column
df_prods_clean=df_prods[df_prods['product_name'].isnull()==False]

In [162]:
# Check the number of rows and columns of df_prods_clean
df_prods_clean.shape

(49677, 5)

# 05 Duplicates

In [163]:
# Create a new df containing duplicate rows from cleaned Products
df_prods_dups=df_prods_clean[df_prods_clean.duplicated()]

In [164]:
df_prods_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [165]:
# Check the number of rows and columns in cleaned Products
df_prods_clean.shape

(49677, 5)

In [166]:
# Remove duplicates from cleaned Products
df_prods_clean_no_dups=df_prods_clean.drop_duplicates()

In [167]:
df_prods_clean_no_dups.shape

(49672, 5)

# 06 Exporting data

In [168]:
# Export Products data with no missings or duplicates
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'products_checked.csv'))

# Task 4.5

# Step 2

In [169]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.214874e+06  
mean         2.776219e+00       1.345202e+01            1.111484e+01  
std          2.046829e+00       4.226088e+00            9.206737e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            4.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

In [170]:
df_ords.head()

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0           0   2539329        1             1                   2   
1           1   2398795        1             2                   3   
2           2    473747        1             3                   3   
3           3   2254736        1             4                   4   
4           4    431534        1             5                   4   

   order_hour_of_day  days_since_prior_order  
0                  8                     NaN  
1                  7                    15.0  
2                 12                    21.0  
3                  7                    29.0  
4                 15                    28.0

In [171]:
df_ords.shape

(3421083, 7)

In [172]:
df_ords['orders_day_of_week'].value_counts(dropna=False)

orders_day_of_week
0    600905
1    587478
2    467260
5    453368
6    448761
3    436972
4    426339
Name: count, dtype: int64

# Descriptive analysis: count of values Count: 
‘days_since_prior_order’ less values than in other columns -> missing values. 
For all other columns the count of values is equal to the number of rows -> no missing data.

# Descriptive analysis. Mean, Median, Quartiles 
‘order_id’: range from 1 to 3.421083e+06. Each row corresponds to one order. 
‘user_id’: range from1 to 2.062090e+05 -> there are 206,209 customers in the database. 
‘order_number’: range from 1 to 100, Q1=5, Q2=11, Q3=23. 25% of customers have made from 1 to 5 orders, and another 25% of customers made from 23 to 100 purchases. And 50% of customers have placed from 5 to 23 orders where the median value is 11 orders vs 17 orders as an average / mean value. So, we have right-skewed distribution. 
‘orders_day_of_week’ : range from 0 to 6. I assume that ‘0’ could be assigned to Sunday and ‘6’ to Saturday. Q1=1, Q2=3 (median), Q3=5. As the median 3 is slightly less than the mean of 2.77, we have left-skewed distribution; a big part of orders is placed on two first days of the week (Sunday and Monday) 
‘order_hour_of_day’ : range from 0 to 23. Q1=1, Q2=13 (median) vs mean value of 13.45, Q3=16. .
‘days_since_prior_order : range from 0 to 30, Q1=4, Q2=7 (median), Q3=15; mean value of 11 is greater than the median, we have right-skewed distribution. 
No deviations are found.

# Step 3. Mixed-typed data in Orders dataframe

In [173]:
# Check for mixed types in Orders
for col in df_ords.columns.tolist() :
    weird=(df_ords[[col]].map(type) !=df_ords[[col]].iloc[0].apply(type)).any(axis=1)
    if len(df_ords[weird])>0:
        print(col)

# No mix-typed data found in Orders dataframe

# Step 5. Missing values in Orders dataframe 

In [174]:
# Check fro missing values in Orders dataframe
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

In [175]:
# Check for null values on days_since_prior_order column
df_ords['days_since_prior_order'].loc[df_ords['days_since_prior_order']==0].value_counts(dropna=False)

days_since_prior_order
0.0    67755
Name: count, dtype: int64

# On ‘days_since_prior_order’, there are 206,209 missing values and 67,755 null values, which is 6.4% and 2.1% of values in this column, respectively. 
The null values can be assigned to the orders when the customer (user) places the order for the first time. In this case null values are relevant. 
NaNs values may be the result of missing data when loading the database from the application. In this case, we need to handle missing values.

In [176]:
# Create subset containing only rows with null 'days_since_prior_order'
df_ords_null=df_ords.loc[df_ords['days_since_prior_order']==0]

In [177]:
df_ords_null

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
8                 8   2295261        1             9                   1   
43               43   2557754        4             5                   5   
189             189   2148135       17            30                   6   
296             296    965160       24            19                   0   
337             337   3162630       27            24                   2   
...             ...       ...      ...           ...                 ...   
3420980     3420980      4277   206206            47                   2   
3420985     3420985    648225   206206            52                   0   
3420991     3420991   2475308   206206            58                   0   
3421001     3421001   1043943   206206            68                   0   
3421044     3421044   2079142   206208            26                   2   

         order_hour_of_day  days_since_prior_order  
8                       16                     0.0  
43                      13                     0.0  
189                     15                     0.0  
296                     16                     0.0  
337                     15                     0.0  
...                    ...                     ...  
3420980                 17                     0.0  
3420985                 19                     0.0  
3420991                 15                     0.0  
3421001                 20                     0.0  
3421044                 16                     0.0  

[67755 rows x 7 columns]

In [178]:
# Count frequency of order_number in df_ords_null
df_ords_null['order_number'].value_counts(dropna=False)

order_number
4      2964
3      2915
2      2899
5      2634
6      2443
       ... 
98      144
100     137
99      134
93      125
97      120
Name: count, Length: 99, dtype: int64

In [179]:
# Check for the first order in the df_ords_null subset
df_ords_null['order_number'].loc[df_ords_null['order_number']==1].value_counts(dropna=False)

Series([], Name: count, dtype: int64)

# There are no orders with number ‘1’ in the subset. My assumption that the null values in ‘days_since_prior_order’ correspond to the first order is wrong.

In [180]:
# Create subset containing only rows with NaNs values in 'days_since_prior_order'
df_ords_nan=df_ords[df_ords['days_since_prior_order'].isnull()==True]

In [181]:
df_ords_nan

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0                 0   2539329        1             1                   2   
11               11   2168274        2             1                   2   
26               26   1374495        3             1                   1   
39               39   3343014        4             1                   6   
45               45   2717275        5             1                   3   
...             ...       ...      ...           ...                 ...   
3420930     3420930    969311   206205             1                   4   
3420934     3420934   3189322   206206             1                   3   
3421002     3421002   2166133   206207             1                   6   
3421019     3421019   2227043   206208             1                   1   
3421069     3421069   3154581   206209             1                   3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 7 columns]

In [182]:
# Count frequency of order_number in df_ords_nan
df_ords_nan['order_number'].value_counts(dropna=False)

order_number
1    206209
Name: count, dtype: int64

# The NaNs values in the column ‘days_since_prior_order’ are assigned to the first order placed by the user.

# The next step is to create a new column indicating the first order

In [183]:
# Step 1: Create a new function to identify the first order
def first_order_function(row): 
    if row['order_number']==1: 
        return True
    else: 
        return False

In [184]:
# Step 2: Apply the new function and check for the result
df_ords.apply(lambda row: first_order_function(row), axis=1)

0           True
1          False
2          False
3          False
4          False
           ...  
3421078    False
3421079    False
3421080    False
3421081    False
3421082    False
Length: 3421083, dtype: bool

In [185]:
# Step 3: Create new column 'first_order'
df_ords['first_order']=df_ords.apply(lambda row: first_order_function(row), axis=1)

In [186]:
df_ords.head()

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0           0   2539329        1             1                   2   
1           1   2398795        1             2                   3   
2           2    473747        1             3                   3   
3           3   2254736        1             4                   4   
4           4    431534        1             5                   4   

   order_hour_of_day  days_since_prior_order  first_order  
0                  8                     NaN         True  
1                  7                    15.0        False  
2                 12                    21.0        False  
3                  7                    29.0        False  
4                 15                    28.0        False

In [187]:
df_ords.shape

(3421083, 8)

In [188]:
df_ords.loc[(df_ords['order_number']==1)&(df_ords['days_since_prior_order'].isnull())&(df_ords['first_order']==True)]

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0                 0   2539329        1             1                   2   
11               11   2168274        2             1                   2   
26               26   1374495        3             1                   1   
39               39   3343014        4             1                   6   
45               45   2717275        5             1                   3   
...             ...       ...      ...           ...                 ...   
3420930     3420930    969311   206205             1                   4   
3420934     3420934   3189322   206206             1                   3   
3421002     3421002   2166133   206207             1                   6   
3421019     3421019   2227043   206208             1                   1   
3421069     3421069   3154581   206209             1                   3   

         order_hour_of_day  days_since_prior_order  first_order  
0                        8                     NaN         True  
11                      11                     NaN         True  
26                      14                     NaN         True  
39                      11                     NaN         True  
45                      12                     NaN         True  
...                    ...                     ...          ...  
3420930                 12                     NaN         True  
3420934                 18                     NaN         True  
3421002                 19                     NaN         True  
3421019                 15                     NaN         True  
3421069                 11                     NaN         True  

[206209 rows x 8 columns]

# As a result, we have a new dataframe with the same number of rows (3421083), with an additional column (totally 8 columns). 'True' value in the first_order column is associated with the '1' value in order_number and Nans in days_since_prior_order  

# Step 7. Duplicates in Orders dataframe

In [189]:
# Check for duplicates 
df_ords_dups=df_ords[df_ords.duplicated()]

In [190]:
df_ords_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order, first_order]
Index: []

In [191]:
df_ords_dups.shape

(0, 8)

# No duplicated rows  are found

# Step 9. Exporting Orders data

In [192]:
# Export Orders data with aditional column 'first_order'
df_ords.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_checked.csv')) 

# Return to Step 1 for consistency check on Products dataframe 

In [193]:
df_prods_clean_no_dups.describe()

product_id      aisle_id  department_id        prices
count  49672.000000  49672.000000   49672.000000  49672.000000
mean   24850.349775     67.762442      11.728942      9.993282
std    14340.705287     38.315784       5.850779    453.615536
min        1.000000      1.000000       1.000000      1.000000
25%    12432.750000     35.000000       7.000000      4.100000
50%    24850.500000     69.000000      13.000000      7.100000
75%    37268.250000    100.000000      17.000000     11.100000
max    49688.000000    134.000000      21.000000  99999.000000

In [194]:
mean_prods=df_prods_clean_no_dups['prices'].mean()

In [195]:
print(mean_prods)

9.993281929457238


In [196]:
std_prods=453.615536

In [197]:
# Calculate Outlier Lower Bound of 'prices'
Lower_bound_prices_prods=mean_prods-std_prods*2

In [198]:
print(Lower_bound_prices_prods)

-897.2377900705428


In [199]:
# Calculate Outlier Upper Bound of 'prices'
Upper_bound_prices_prods=mean_prods+std_prods*2

In [200]:
print(Upper_bound_prices_prods)

917.2243539294573


# Given the common rule the data points more than 2 standard deviations away from the mean are outliers for normal distribution data, any prices greater than 917.224354 are considered outliers   

In [201]:
# Check for outliers in 'prices'
df_prods_clean_no_dups.loc[df_prods_clean_no_dups['prices']>Upper_bound_prices_prods]

product_id                      product_name  aisle_id  department_id  \
21554       21553  Lowfat 2% Milkfat Cottage Cheese       108             16   
33666       33664             2 % Reduced Fat  Milk        84             16   

        prices  
21554  14900.0  
33666  99999.0

# There are 2 ourliers. I suggest removing the rows with outliers

In [202]:
# Check for rows and columns in Products 
df_prods_clean_no_dups.shape

(49672, 5)

In [203]:
# Remove outliers and create a new Products dataframe
df_prods_clean_no_dups_no_outliers=df_prods_clean_no_dups[(df_prods_clean_no_dups['prices']>Upper_bound_prices_prods)==False]

In [204]:
df_prods_clean_no_dups_no_outliers.shape

(49670, 5)

In [205]:
df_prods_clean_no_dups_no_outliers.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [206]:
df_prods_clean_no_dups_no_outliers.describe()

product_id      aisle_id  department_id        prices
count  49670.000000  49670.000000   49670.000000  49670.000000
mean   24850.238716     67.761305      11.728770      7.680437
std    14340.931850     38.316060       5.850834      4.199381
min        1.000000      1.000000       1.000000      1.000000
25%    12432.250000     35.000000       7.000000      4.100000
50%    24850.500000     69.000000      13.000000      7.100000
75%    37268.750000    100.000000      17.000000     11.100000
max    49688.000000    134.000000      21.000000     25.000000

In [207]:
# Exporting Products data with no missings, duplicates or outliers and repalce the previous expoted csv file 'products_checked'
df_prods_clean_no_dups_no_outliers.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'products_checked.csv'))